nhs/data/join_on_sa1.py

```diff
@@ -0,0 +1,219 @@
+    """Parallel process point data."""
+    chunks = np.array_split(pnts_gdf, num_cores)
+    with Pool(num_cores) as pool:
+        all_results = pool.map(lambda chunk: process_chunk(map_data, chunk), chunks)
+    return pl.concat(all_results)
+
+
+def random_distribution(df: pl.DataFrame, num_iterations: int = 10) -> pl.DataFrame:
+    """Randomly shuffle lat-long pairs within their respective areas."""
+    model_df = df.clone()
+    model_df = model_df.with_columns(
+        [pl.col("y").alias("original_lat"), pl.col("x").alias("original_lon")]
+    )
+
+    for _ in range(num_iterations):
+        # Loop over unique area_code values and shuffle lat-long pairs within each group
+        area_codes = model_df["area_code"].unique().to_list()
+        for area_code in area_codes:
+            area_df = model_df.filter(pl.col("area_code") == area_code)
+            shuffled_pairs = area_df[["x", "y"]].sample(frac=1).to_dict(as_series=False)
+            model_df = model_df.with_columns(
+                [
+                    pl.when(pl.col("area_code") == area_code)
+                    .then(pl.Series(shuffled_pairs["x"]))
+                    .otherwise(pl.col("x"))
+                    .alias("x"),
+                    pl.when(pl.col("area_code") == area_code)
+                    .then(pl.Series(shuffled_pairs["y"]))
+                    .otherwise(pl.col("y"))
+                    .alias("y"),
+                ]
+            )
+
+    return model_df
+
+
+def visualize_results(final_result_df: pl.DataFrame, area_codes: List[str]) -> None:
+    """Visualize the original and final positions of points."""
+    for area_code in area_codes:
+        area_result_df = final_result_df.filter(pl.col("area_code") == area_code)
+
+        plt.figure(figsize=(12, 6))
+        plt.subplot(1, 2, 1)
+        plt.scatter(
+            area_result_df["original_lon"],
+            area_result_df["original_lat"],
+            c=area_result_df["attribute"].to_numpy(),
+            cmap="tab10",
+            s=50,
+            alpha=0.7,
+        )
+        plt.title(f"Original Positions for Area Code: {area_code}")
+        plt.xlabel("Original Longitude")
+        plt.ylabel("Original Latitude")
+
+        plt.subplot(1, 2, 2)
+        plt.scatter(
+            area_result_df["x"],
+            area_result_df["y"],
+            c=area_result_df["attribute"].to_numpy(),
+            cmap="tab10",
+            s=50,
+            alpha=0.7,
+        )
+        plt.title(
+            f"Final Positions after Random Distribution for Area Code: {area_code}"
+        )
+        plt.xlabel("Final Longitude")
+        plt.ylabel("Final Latitude")
+
+        plt.tight_layout()
+        plt.show()
+
+
+def main() -> None:
+    config_path = "configurations.yaml"
+    config = read_config(config_path)
+
+    shapefile_path = config["shapefile_path"]
+    csv_path = config["csv_path"]
+    crs = config["crs"]
+
+    total_rows = read_shapefile(shapefile_path)
```

Comment:
I'm not sure if the shapefiles can be parquet-ed, if we can read it as a `DataFrame` or `LazyFrame` then yes it can be read as parquet file. Only relevant if loading the shape files take a very long time

Reply:
Shapefiles can be parqueted if you buy a certain subscription to a certain application, but otherwise I don't think so. Shapefile loading takes ~1min iirc.